mlfinlab/feature_importance/cfi.py

```diff
@@ -0,0 +1,126 @@
+"""
+An Implementation of the Clustered Feature Importance algorithm described by
+Dr Marcos Lopez de Prado in 'Clustered Feature Importance (Presentation Slides)'
+(https://papers.ssrn.com/sol3/papers.cfm?abstract_id=3517595)
+"""
+
+#Imports
+import numpy as np
+import pandas as pd
+import mlfinlab as ml
+from sklearn.metrics import log_loss
+from scipy.cluster.hierarchy import linkage, fcluster
+from mlfinlab.codependence import variation_of_information_score
+
+def get_VI_matrix(df : pd.DataFrame):
+    '''
+    Get Variation of Information (VI) matrix.
+
+    :param df: (pd.DataFrame) of features
+    :return: (pd.DataFrame) of VI matrix
+    '''
+    VI = pd.DataFrame(index=df.columns)
+    features_cols = df.columns.values
+    for col0 in features_cols:
+        vi = []
+        for col1 in features_cols:
+            x = df[col0].values
+            y = df[col1].values
+            vi.append(variation_of_information_score(x,y,n_bins=None,normalize=True))
+        VI[col0] = vi
+    return VI
+
+def cluster_features(X : pd.DataFrame, C : int = None):
+    '''
+    Get Clusters containing features subsets
+
+    (param) X : (pd.DataFrame) of features
+    (param) C : (int) number of clusters to form.
+                      Must be less the total number of features.
+                      If None then it returns optimal number of clusters
+                      decided by the ONC Algorithm.
+
+    (return) : (array) of feature subsets
+    '''
+
+    vi_matrix =  get_VI_matrix(X)
+    link = linkage(vi_matrix, 'single')
```

Comment:
As far as I understand we should apply linkage to distance matrix rather than dependence. So in this case we calculate dependence matrix  (vi_matrix) -> distance, for example angular (or abs angular or squared angular) can also be a function parameter on dependence matrix -> linkage(dependence matrix)

Reply:
I get it but the distance matrix can be applied during the presence of linear substitution effect, but the VI matrix is effective in presence of both linear and non-linear substitution effects. I really want to know your @proskurin opinion on these.

Reply:
As far as I understand distance matrix is always an input for clustering. If for example we have negative values (for linear correlation for example), clustering algorithm will raise error. So VI is a measure of dependence which always needs to be transformed  into distance matrix. As far as I remember, Marcos stated that we need to apply angular distance operator before clustering

Reply:
Okay, as you mentioned to add a universal function (in the above comments), I would create a new script in the codependence section that can get both dependence and distance matrix.